plugins/org.eclipse.fordiac.ide.library/src/org/eclipse/fordiac/ide/library/LibraryManager.java

```diff
@@ -596,72 +599,84 @@ public void checkManifestFile(final IProject project, final TypeLibrary typeLibr
 		}
+		startLocalResolveJob(project, typeLibrary);
+
 	}
 
 	/**
 	 * Check if library dependency is present in the {@link IProject} and
 	 * download/import as needed if not
 	 *
-	 * @param req         dependency
-	 * @param projectLibs libraries already present in project
+	 * @param req         library dependency
 	 * @param project     selected project
 	 * @param typeLibrary {@link TypeLibrary} to use
 	 */
-	private void checkLibrary(final Required req, final Map<String, List<String>> projectLibs, final IProject project,
-			final TypeLibrary typeLibrary) {
+	private void checkDependency(final Required req, final IProject project, final TypeLibrary typeLibrary) {
 		checkLibChanges();
+		String preferred = null;
+
 		// check already linked lib
-		if (projectLibs.containsKey(req.getSymbolicName()) && (projectLibs.get(req.getSymbolicName()).stream()
-				.filter(p -> versionComparator.contains(req.getVersion(), p)).count() > 0)) {
-			return;
+		final IFolder libFolder = project.getFolder(TYPE_LIB_FOLDER_NAME).getFolder(req.getSymbolicName());
+		if (libFolder.exists()) {
+			final Manifest libManifest = ManifestHelper.getContainerManifest(libFolder);
+			if (libManifest != null) {
+				if (VersionComparator.contains(req.getVersion(),
+						libManifest.getProduct().getVersionInfo().getVersion())) {
+					return;
+				}
+			} else {
+				final IPath path = libFolder.getLocation();
+				final String segment = (path.segmentCount() >= 2) ? path.segment(path.segmentCount() - 2) : ""; //$NON-NLS-1$
+				final int index = segment.lastIndexOf('-');
+				if (index > 0) {
+					preferred = segment.substring(index + 1);
+					if (!VersionComparator.contains(req.getVersion(), preferred)) {
+						preferred = null;
+					}
+				}
+			}
 		}
 
-		// check standard libs
-		if (stdlibraries.containsKey(req.getSymbolicName()) && stdlibraries.get(req.getSymbolicName()).stream()
-				.anyMatch(l -> versionComparator.parseVersionRange(req.getVersion()).includes(l.version()))) {
-
-			startLocalLinkJob(req, project, stdlibraries, typeLibrary);
-			return;
-		}
+		if (preferred != null) {
+			final Required prefReq = ManifestHelper.createRequired(req.getSymbolicName(), preferred);
+			// check standard libs
+			if (checkLibraries(prefReq, project, typeLibrary, stdlibraries)) {
+				return;
+			}
 
-		// check local libs
-		if (libraries.containsKey(req.getSymbolicName()) && libraries.get(req.getSymbolicName()).stream()
-				.anyMatch(l -> versionComparator.parseVersionRange(req.getVersion()).includes(l.version()))) {
+			// check local libs
+			if (checkLibraries(prefReq, project, typeLibrary, libraries)) {
+				return;
+			}
+		} else {
+			// check standard libs
+			if (checkLibraries(req, project, typeLibrary, stdlibraries)) {
+				return;
+			}
 
-			startLocalLinkJob(req, project, libraries, typeLibrary);
-			return;
+			// check local libs
+			if (checkLibraries(req, project, typeLibrary, libraries)) {
+				return;
+			}
```

Comment:
This code is hard to read.
For example you can make methods, checkStandardLib, checkLocalLib, as you use that multiple times